File: src/EmployeeOfTheMonth/EmployeeOfTheMonth/Model/WorkHour.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Text;

namespace EmployeeOfTheMonth.Model
{
    class WorkHour
    {
        public DateTime Entrance { get; set; }
        public DateTime StartLunch { get; set; }
        public DateTime EndLunch { get; set; }
        public DateTime Exit { get; set; }

        public DateTime EntranceExtra { get; set; }
        public DateTime ExitExtra { get; set; }

        public TimeSpan TotalHours
        {
            get { return Exit.Subtract(Entrance) - EndLunch.Subtract(StartLunch); }
        }

        public TimeSpan TotalLunch
        {
            get { return EndLunch.Subtract(StartLunch); }
        }


        private DateTime _base;
        private readonly Random _random;
        private int GetMinutes()
        {
            return _random.Next(_base.Day, _base.Day + 30);
        }
        public WorkHour(DateTime date)
        {
            _base = date;
            _random = new Random(date.Day + Environment.MachineName.GetHashCode() + Environment.TickCount);
            Entrance = date.AddHours(Constants.MinimumEntrance).AddMinutes(GetMinutes());
            StartLunch = Entrance.AddHours(Constants.MinimumHoursBetweenEntranceAndLunchStart).AddMinutes(GetMinutes());
            EndLunch = StartLunch.AddHours(Constants.MinimumLunchHours).AddMinutes(GetMinutes());
            Exit = Entrance.AddHours(Constants.TotalWorkHours).Add(TotalLunch).AddMinutes(GetMinutes());
        }

        public override string ToString()
        {
            var cultureInfo = new CultureInfo("pt-BR");
            if (_base.DayOfWeek == DayOfWeek.Saturday || _base.DayOfWeek == DayOfWeek.Sunday)
                return string.Format("{0:D2}   | {1}        | *****   | *****         | *****          | *****", _base.Day, _base.ToString("ddd", cultureInfo));
            return string.Format("{0:D2}   | {1}        | {2:HH:mm}   | {3:HH:mm}         | {4:HH:mm}          | {5:HH:mm}", _base.Day,
                                 _base.ToString("ddd", cultureInfo), Entrance, StartLunch, EndLunch, Exit);
        }

    }
}

File: js/workhour.js
function WorkHour(day) {

	var self = this;

	var year = parseInt($("#year").val());
	var month = parseInt($("#month").val()) + 1;
	

	var date = new Date(year, month, day);
	var week = ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"];

	// Contants

	this.MinimumEntrance = 9;
	this.MinimumHoursBetweenEntranceAndLunchStart = 3;
	this.MinimumLunchHours = 1;
	this.TotalWorkHours = 8;

	this.getRandomMinutes = function() { return Math.floor(Math.random() * 22); };

	this.TotalLunch = function() { return (new Date(this.EndLunch) - new Date(this.StartLunch)) / 60000; };

	this.padLeftMinutes = function(minutes) { 
		if (!minutes || minutes.toString().length == 0) return "00"; 
		if (minutes.toString().length == 1) return "0" + minutes;
		return minutes;
	}

	this.Entrance = date.addHours(this.MinimumEntrance).addMinutes(this.getRandomMinutes()); 
	this.StartLunch = new Date(this.Entrance).addHours(this.MinimumHoursBetweenEntranceAndLunchStart).addMinutes(this.getRandomMinutes()); 
	this.EndLunch = new Date(this.StartLunch).addHours(this.MinimumLunchHours).addMinutes(this.getRandomMinutes()); 
	this.Exit = new Date(this.Entrance).addHours(this.TotalWorkHours).addMinutes(this.TotalLunch());

	if (!$("#exact8").hasClass("btn-success"))
		if (day % 2 == 0)
			this.Exit = new Date(this.Exit).addMinutes(this.getRandomMinutes() * -1);
		else
			this.Exit = new Date(this.Exit).addMinutes(this.getRandomMinutes());

	function renderWeekend(weekday) {
		var tr = $("<tr />");
		var td = $("<td />");

		tr
			.append(td.clone().text((day) + "/" + month))
			.append(td.clone().text(week[weekday]))
			.append(td.clone().text("**:**"))
			.append(td.clone().text("**:**"))
			.append(td.clone().text("**:**"))
			.append(td.clone().text("**:**"));
			
		return tr;
	}

	this.RenderAsRow = function() {
		var tr = $("<tr />");
		var td = $("<td />");

		var weekday = new Date(year, month - 1, day).getDay();
		if (weekday == 0 || weekday == 6) return renderWeekend(weekday);

		tr
			.append(td.clone().text((day) + "/" + month))
			.append(td.clone().text(week[weekday]))
			.append(td.clone().text(new Date(self.Entrance).getHours() + ":" + self.padLeftMinutes(new Date(self.Entrance).getMinutes())))
			.append(td.clone().text(new Date(self.StartLunch).getHours() + ":" + self.padLeftMinutes(new Date(self.StartLunch).getMinutes())))
			.append(td.clone().text(new Date(self.EndLunch).getHours() + ":" + self.padLeftMinutes(new Date(self.EndLunch).getMinutes())))
			.append(td.clone().text(new Date(self.Exit).getHours() + ":" + self.padLeftMinutes(new Date(self.Exit).getMinutes())));

		return tr;

	};



	return this;
}